File: Python/CreationalDesignPattern/FacadePatternDemo.py
class TV:
    def on(self):
        print("TV is ON")

    def set_input(self):
        print("TV input set to HDMI")


class SoundSystem:
    def on(self):
        print("Sound System is ON")

    def set_volume(self, volume: int):
        print(f"Volume set to {volume}")


class StreamingDevice:
    def on(self):
        print("Streaming Device is ON")

    def play_movie(self, movie: str):
        print(f"Playing movie: {movie}")


class HomeTheaterFacade:
    def __init__(self):
        self.tv = TV()
        self.sound_system = SoundSystem()
        self.streaming_device = StreamingDevice()

    def watch_movie(self, movie: str):
        print("Preparing Home Theater...\n")

        self.tv.on()
        self.tv.set_input()

        self.sound_system.on()
        self.sound_system.set_volume(20)

        self.streaming_device.on()
        self.streaming_device.play_movie(movie)

        print("\nEnjoy your movie!")


def main():
    home_theater = HomeTheaterFacade()
    home_theater.watch_movie("Interstellar")


if __name__ == "__main__":
    main()

File: C++/CreationalDesignPattern/FacadePatternDemo.cpp
#include <iostream>
#include <string>

// Subsystem 1
class TV {
public:
    void on() { std::cout << "TV is ON" << std::endl; }

    void setInput() { std::cout << "TV input set to HDMI" << std::endl; }
};

// Subsystem 2
class SoundSystem {
public:
    void on() { std::cout << "Sound System is ON" << std::endl; }

    void setVolume(int volume) { std::cout << "Volume set to " << volume << std::endl; }
};

// Subsystem 3
class StreamingDevice {
public:
    void on() { std::cout << "Streaming Device is ON" << std::endl; }

    void playMovie(const std::string& movie) {
        std::cout << "Playing movie: " << movie << std::endl;
    }
};

// Facade
class HomeTheaterFacade {
    TV tv;
    SoundSystem soundSystem;
    StreamingDevice streamingDevice;

public:
    void watchMovie(const std::string& movie) {
        std::cout << "Preparing Home Theater...\n" << std::endl;

        tv.on();
        tv.setInput();

        soundSystem.on();
        soundSystem.setVolume(20);

        streamingDevice.on();
        streamingDevice.playMovie(movie);

        std::cout << "\nEnjoy your movie!" << std::endl;
    }
};

int main() {
    HomeTheaterFacade homeTheater;
    homeTheater.watchMovie("Interstellar");
    return 0;
}
